### ISSUE TITLE
Add workflow to check project metadata

### ISSUE THREAD

**Issue by andrewtavis** [scribe-org/Scribe-Data#340]
### Terms

- [X] I have searched [open and closed feature requests](https://github.com/scribe-org/Scribe-Data/issues?q=is%3Aissue+label%3Afeature)
- [X] I agree to follow Scribe-Data's [Code of Conduct](https://github.com/scribe-org/Scribe-Data/blob/main/.github/CODE_OF_CONDUCT.md)

### Description

This issue would create a new workflow in [.github/workflows](https://github.com/scribe-org/Scribe-Data/tree/main/.github/workflows) called `check_project_metadata.yaml` that would call Python scripts that would check the project's metadata files `language_metadata.json` and `data_type_metadata.json`. We can put these scripts in a new [.github/workflows](https://github.com/scribe-org/Scribe-Data/tree/main/src/scribe_data) directory called `check`. The scripts would be:

- `/src/scribe_data/check/check_language_metadata.py` would check the `language_data_extraction` directory and make sure that all languages in this directory are included in the `language_metadata.json` file
    - Note: We need #293 to be finished before this one is made

A code snippet for this comes from #330:

```py
def get_available_languages() -> list[tuple[str, str]]:
    """
    Get available languages from the data extraction folder.

    Returns
    -------
        list[tuple[str, str]]: A list of tuples with the language name and its QID.
    """
    extraction_dir = LANGUAGE_DATA_EXTRACTION_DIR
    available_languages = []
    for lang_folder in extraction_dir.iterdir():
        if lang_folder.is_dir():  # Check if it's a directory
            lang_name = lang_folder.name
```

- `/src/scribe_data/check/check_data_type_metadata.py` would check each sub directory of the language directories and assure that the included sub directories that are named for data types are also included in the `data_type_metadata.json` file
- Both scripts need to account for the fact that we have some languages that are sub-languages of a meta one like `Norwrgian` or `Hindustani`
- We also need both of these scripts to confirm that there are no entries in the JSONs that are not in the directories, so say that we move a language to be a sub-language, then this needs to be reflected

### Contribution

Happy to support, answer questions and review as needed!

CC @DeleMike and @catreedle :)

**Comment by KesharwaniArpita** [comment]
Hi, @andrewtavis , @DeleMike and @catreedle, Can I also contribute to this issue?

**Comment by DeleMike** [comment]
Nice @KesharwaniArpita , I think you can. 

Could you wait a bit for feedback from @catreedle , I wanna hear her thoughts :)

**Comment by KesharwaniArpita** [comment]
Yeah Sure!!

**Comment by DeleMike** [comment]
Thank you!

**Comment by catreedle** [comment]
> Nice @KesharwaniArpita , I think you can.
> 
> Could you wait a bit for feedback from @catreedle , I wanna hear her thoughts :)

Sure, @DeleMike @KesharwaniArpita! I'd also be happy to help in any way I can.

**Comment by catreedle** [comment]
I'm thinking we could divide the task into:
- check_language_metada
- check_data_type_metadata
- configuring GitHub workflow and adding unit tests

What do you guys think? @DeleMike @KesharwaniArpita

**Comment by KesharwaniArpita** [comment]
Works for me!! Thanks @catreedle

**Comment by DeleMike** [comment]
looking at this now...

**Comment by DeleMike** [comment]
@catreedle, I like the way you divided the tasks! We could make these three(3) PRs for each issue. However, `check_language_metadata` check and `check_data_type_metadata` has to be resolved before we can even write tests and finish up the GH workflow. Nevertheless, the initial GH workflow setup can be started.

What do you think @KesharwaniArpita  and  @andrewtavis ?

**Comment by DeleMike** [comment]
Also, @andrewtavis , do we hold off on this as #293 has not been finished, yeah?

**Comment by andrewtavis** [comment]
Let's check with @OmarAI2003 on what the estimated time for #293 is :) I think that you all can at least start the work here as only the script relies on #293, and minor edits can be done to the PR once the work in the other issue is done 😊

**Comment by catreedle** [comment]
I'd like to work on check_language_metadata. Would that be alright?

**Comment by DeleMike** [comment]
> Let's check with @OmarAI2003 on what the estimated time for #293 is :) I think that you all can at least start the work here as only the script relies on #293, and minor edits can be done to the PR once the work in the other issue is done 😊

cool then!

**Comment by DeleMike** [comment]
> I'd like to work on check_language_metadata. Would that be alright?

yes, that's fine. @KesharwaniArpita can you take `check_data_type_metadata` while I take the last one?

**Comment by KesharwaniArpita** [comment]
Yeah sure! I would love to @DeleMike

**Comment by DeleMike** [comment]
> Yeah sure! I would love to @DeleMike

great, thanks. so everyone's settled! 🙌🏾

**Comment by catreedle** [comment]
Hi @andrewtavis, I want to ask how would the JSON structure for language with sub directory look like?
like any other language
```
{
  "language": "nynorsk",
  "iso": "es",
  "qid": "Q1321",
  "remove-words": ["of", "the", "The", "and"],
  "ignore-words": []
},
{
  "language": "bokmal",
  "iso": "sv",
  "qid": "Q9027",
  "remove-words": ["of", "the", "The", "and", "Checklist", "Catalogue"],
  "ignore-words": ["databasdump"]
}

```

with special key

```
{
  "Norwegian": {
    "languages": [
      {
        "language": "nynorsk",
        "iso": "es",
        "qid": "Q1321",
        "remove-words": ["of", "the", "The", "and"],
        "ignore-words": []
      },
      {
        "language": "bokmal",
        "iso": "sv",
        "qid": "Q9027",
        "remove-words": ["of", "the", "The", "and", "Checklist", "Catalogue"],
        "ignore-words": ["databasdump"]
      }
    ]
  }
}

```

or something else?

**Comment by andrewtavis** [comment]
@OmarAI2003 and I agreed on something like this:

```json
{
  "english": {
    "iso": "en",
    "qid": "Q1860"
  },
  "norwegian": {
    "dialects": { 
      "nynorsk": {
        "iso": "nn",
        "qid": "Q103377"
      },
      "bokmål": {
        "iso": "nb",
        "qid": "Q34770"
      }
    }
}
```

We're getting rid of the remove words and ignore words. Won't be used anymore :)

**Comment by andrewtavis** [comment]
To add syntax highlighting to your code snippets you can add the language after the first three backticks, @catreedle :) Just so it's a bit more readable.

**Comment by OmarAI2003** [comment]
> @OmarAI2003 and I agreed on something like this:
> 
> ```json
> {
>   "english": {
>     "iso": "en",
>     "qid": "Q1860"
>   },
>   "norwegian": {
>     "dialects": { 
>       "nynorsk": {
>         "iso": "nn",
>         "qid": "Q103377"
>       },
>       "bokmål": {
>         "iso": "nb",
>         "qid": "Q34770"
>       }
>     }
> }
> ```
> 
> We're getting rid of the remove words and ignore words. Won't be used anymore :)

This is pretty much correct, except for the part where dialects are called sub-languages.

**Comment by catreedle** [comment]
noted. thank you @OmarAI2003 @andrewtavis :)

**Comment by catreedle** [comment]
> To add syntax highlighting to your code snippets you can add the language after the first three backticks, @catreedle :) Just so it's a bit more readable.

good to know, thanks! 😊

**Comment by andrewtavis** [comment]
I would structure it as a dictionary and then have them in sub dictionaries so we can mimic what the functionality will be with the meta data file when it's finished :) Ideally we'd just load that file in and the check would follow directly from it.

**Comment by catreedle** [comment]
Ok, thanks!

**Comment by KesharwaniArpita** [comment]
Hi @DeleMike , @catreedle and @andrewtavis . This is what I understand :
We are creating a new workflow in `.github/workflows` named `check_project_metadata.yaml`, which will call two Python scripts to validate the project's metadata files: `language_metadata.json` and `data_type_metadata.json`.  For my part I need to developed the `check_data_type_metadata.py` script, which ensures consistency between the language data type directories and the `data_type_metadata.json file` in the project. The script traverses the directory structure under `src/scribe_data/language_data_extraction/"LANGUAGE"`, checking for data type subdirectories and accounting for meta-languages and their sub-languages (e.g., `Hindi` and `Urdu` under `Hindustani`). It verifies that all data types present in the directories are also reflected in the JSON file and flags any discrepancies, such as missing or extra data types. This functionality will help us maintain synchronization between our directory structure and metadata.
And the script needs to be created under a new directory `/src/scribe_data/check/check_data_type_metadata.py`. This automation helps ensure that all languages and their corresponding data types are accurately represented in the directories and metadata files, preventing discrepancies that could arise from manual updates.

Is this alright?

**Comment by catreedle** [comment]
Hi @KesharwaniArpita,
Looks good to me! I think it covers everything, even the part where there might be extra data types that slipped my mind. :)

For my part, I’m tracking languages that are in the directory but missing from `language_metadata.json`, as well as those in `language_metadata.json` but not in the directory. I'm keeping both as a separate list and logging them separately.
Can I approach it this way, @andrewtavis?

**Comment by KesharwaniArpita** [comment]
@catreedle How do you plan to deal with the grouped languages.

**Comment by andrewtavis** [comment]
This sounds good, @catreedle :) We should definitely be tracking both 😊

**Comment by DeleMike** [comment]
> Hi @DeleMike , @catreedle and @andrewtavis . This is what I understand : We are creating a new workflow in `.github/workflows` named `check_project_metadata.yaml`, which will call two Python scripts to validate the project's metadata files: `language_metadata.json` and `data_type_metadata.json`. For my part I need to developed the `check_data_type_metadata.py` script, which ensures consistency between the language data type directories and the `data_type_metadata.json file` in the project. The script traverses the directory structure under `src/scribe_data/language_data_extraction/"LANGUAGE"`, checking for data type subdirectories and accounting for meta-languages and their sub-languages (e.g., `Hindi` and `Urdu` under `Hindustani`). It verifies that all data types present in the directories are also reflected in the JSON file and flags any discrepancies, such as missing or extra data types. This functionality will help us maintain synchronization between our directory structure and metadata. And the script needs to be created under a new directory `/src/scribe_data/check/check_data_type_metadata.py`. This automation helps ensure that all languages and their corresponding data types are accurately represented in the directories and metadata files, preventing discrepancies that could arise from manual updates.
> 
> Is this alright?

Yes @KesharwaniArpita, you are right! and this sounds good to me!

**Comment by catreedle** [comment]
> How do you plan to deal with the grouped languages

For the languages from the directory, I use a dictionary where the parent language is the key, and the sub-languages are stored under a `sub-languages` key. I use a flag to determine if the sub-directories of a parent folder contain data types (by checking for keys in `data_type_metadata.json`). If they do, I conclude that the folder doesn’t have sub-languages. otherwise, it does.

However, I just realized there could be a problem when a folder structure looks like this:

--Indonesian
----verbs
----Javanese
----Sundanese

If "verbs" is mistakenly included, "Javanese" and "Sundanese" won't be recognized as sub-languages. I’ll be working on a fix for this! 😅

**Comment by catreedle** [comment]
opened a PR [here](https://github.com/scribe-org/Scribe-Data/pull/385), feedback will be much appreciated 😊

**Comment by DeleMike** [comment]
drop a few comments 😄 @catreedle 
If you have questions, I'm a comment away 😄

**Comment by catreedle** [comment]
Thank you! @DeleMike Really appreciate it 😊

**Comment by andrewtavis** [comment]
Closed by the above PRs 🚀 Thanks all for the amazing work!

We should look into simplifying all this down to maybe three checks:

- `check_project_metadata` makes sure the metadata files are correct
- `check_project_structure` makes sure the structure is valid given the files
- `check_project_queries` the combination of `check_query_identifiers` and `check_query_forms`

Let me know how this sounds and we can make another issue!

**Comment by KesharwaniArpita** [comment]
Makes Sense to me @andrewtavis.
